File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode head;
        ListNode cur;
        if (list1 == null) {
            return list2;
        } else if (list2 == null) {
            return list1;
        }
        
        if (list1.val < list2.val) {
            head = list1;
            list1 = list1.next;
        } else {
            head = list2;
            list2 = list2.next;
        }

        cur = head;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                ListNode temp = list1.next;
                cur.next = list1;
                cur = cur.next;
                list1 = temp;
            } else {
                ListNode temp = list2.next;
                cur.next = list2;
                cur = cur.next;
                list2 = temp;
            }
        }

        if (list1 == null) {
            while (list2 != null) {
                ListNode temp = list2.next;
                cur.next = list2;
                cur = cur.next;
                list2 = temp;
            }
        } else if (list2 == null) {
            while (list1 != null) {
                ListNode temp = list1.next;
                cur.next = list1;
                cur = cur.next;
                list1 = temp;
            }
        }

        return head;

    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        left = head
        right = head
        cnt = 1
        firstTime = True
        newHead = None
        prev = None

        while right:
            if cnt == k:
                node = None
                tail = left
                
                for i in range(k):
                    temp = left.next
                    left.next = node
                    node = left
                    left = temp
                tail.next = left
                if firstTime:
                    firstTime = False
                    newHead = right
                else:
                    prev.next = right
                prev = tail
                right = left
                cnt = 1
            else:
                right = right.next
                cnt += 1
        return newHead
